source/s3_auto_ranged_get.c

```diff
@@ -325,97 +401,195 @@ static int s_s3_auto_ranged_get_prepare_request(
     return AWS_OP_ERR;
 }
 
-static void s_s3_auto_ranged_get_request_finished(
+/* Using the response returned from the given request, determine the overall object-range and content-length. */
+static int s_discover_object_range_and_content_length(
     struct aws_s3_meta_request *meta_request,
     struct aws_s3_request *request,
-    int error_code) {
-    AWS_PRECONDITION(meta_request);
-    AWS_PRECONDITION(meta_request->impl);
-    AWS_PRECONDITION(request);
+    int error_code,
+    uint64_t *out_total_content_length,
+    uint64_t *out_object_range_start,
+    uint64_t *out_object_range_end,
+    uint64_t *out_total_object_size) {
+    AWS_PRECONDITION(out_total_content_length);
+    AWS_PRECONDITION(out_object_range_start);
+    AWS_PRECONDITION(out_object_range_end);
+    AWS_PRECONDITION(out_total_object_size);
+
+    int result = AWS_OP_ERR;
+
+    uint64_t total_content_length = 0;
+    uint64_t object_range_start = 0;
+    uint64_t object_range_end = 0;
+    uint64_t total_object_size = 0;
+
+    AWS_ASSERT(request->discovers_object_size);
+
+    switch (request->request_tag) {
+        case AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_HEAD_OBJECT:
+            if (error_code != AWS_ERROR_SUCCESS) {
+                /* If the head request failed, there's nothing we can do, so resurface the error code. */
+                aws_raise_error(error_code);
+                break;
+            }
 
-    (void)request;
-    (void)error_code;
+            /* There should be a Content-Length header that indicates the total size of the range.*/
+            if (aws_s3_parse_content_length_response_header(
+                    meta_request->allocator, request->send_data.response_headers, &total_content_length)) {
 
-    struct aws_s3_auto_ranged_get *auto_ranged_get = meta_request->impl;
+                AWS_LOGF_ERROR(
+                    AWS_LS_S3_META_REQUEST,
+                    "id=%p Could not find content-length header for request %p",
+                    (void *)meta_request,
+                    (void *)request);
+                break;
+            }
 
-    uint32_t num_parts = 0;
+            /* When doing a head object request, we currently assume that we did so with a message that had a Range
+             * header. This means that there should also be a Content-Range header that specifies the object range and
+             * total object size.*/
+            if (aws_s3_parse_content_range_response_header(
+                    meta_request->allocator,
+                    request->send_data.response_headers,
+                    &object_range_start,
+                    &object_range_end,
+                    &total_object_size)) {
```

Comment:
S3 should return an error in that case, causing the meta request to die.  We can add a test for that.

For the range not matching, I think the service should own that behavior.  I think there are valid cases where the size could be different that what they asked for.  (For example, S3 will return the part of a range it can satisfy, even if it's smaller or goes beyond the end of the file.)

Reply:
Added the test!

Reply:
Yeah, it makes sense. Thanks!